some lgtm code consolidation

jlaehne opened the pull request:
I started to fix some of the issues identified by [lgtm](https://lgtm.com/projects/g/hyperspy/hyperspy), see also #2966 

## Progress of the PR:

- [x] changes implemented
- [x] ready for review

ericpre committed:
Update changelog - run `towncrier build`

jlaehne committed:
some lgtm code consolidation

jlaehne committed:
fix errors

jlaehne committed:
some more lgtm fixes

jlaehne commented:
I won't find more time to continue for the moment

jlaehne committed:
some more lgtm fixes

ericpre committed:
Fix formatting changelog

ericpre commented:
ok, let's put it in draft for now, then.

ericpre reviewed (commented):
It would be good to add the corresponding documentation to the dev guide to help contributors finding relevant documentation on lgtm and briefly explain what are the comments used by lgtm, etc.

jlaehne committed:
more lgtm

jlaehne committed:
remove duplicate imports

jlaehne committed:
typos

jlaehne committed:
duplicate assignment

jlaehne commented:
Added some more corrections based on the checks by lgtm.com

Fixes a number of real issues and does some streamlining, such as removing duplicate imports.

I'm loosing the overview which issues I have addressed and which not.

Also, I think if adding more changes, it will get hard to review.

Therefore, I submit this for review and propose to continue in a separate PR once lgtm.com has reevaluated the HyperSpy.

ericpre commented on hyperspy/_components/eels_cl_edge.py:
```diff
@@ -211,7 +211,7 @@ def _calculate_effective_angle(self):
                 self.GOS.onset_energy,
                 self.convergence_angle,
                 self.collection_angle)
-        except BaseException:
```
If my understanding is correct, the difference between the two is that `BaseException` will catch slightly more exception (`SystemExit`, `KeyboardInterrupt`, `GeneratorExit`) than `Exception` (from https://docs.python.org/3/library/exceptions.html#exception-hierarchy), but I don't know the reason to favour the use of one of the two.
@francisco-dlp, if I remember correctly, I have seen you using `BaseException`, do you have a view on this?

jlaehne replied:
According to lgtm, that is the point. `BaseException` is very general: https://lgtm.com/rules/6780080/

ericpre commented on hyperspy/_signals/lazy.py:
```diff
@@ -23,7 +23,8 @@
 import dask.array as da
-import dask
+import dask.threaded as dt
+from dask import __version__
```
Is it good? This will add the `__version__` attribute to this module with its value set to the dask version!

jlaehne replied:
It avoids importing the whole of dask, just to have the version available. Any better idea?

ericpre commented on hyperspy/signal.py:
```diff
@@ -39,12 +39,12 @@
 
-import hyperspy
+import hyperspy.io as hypio
+import hyperspy.misc.array_tools as hypat
```
This is subjective but I am not sure if this make the code better in a sense, as in more readable.

jlaehne replied:
The file imports a dozen different modules from hyperspy independently, but then the whole of hyperspy again just for these two submodules.

ericpre reviewed (commented):
This is useful but my impression is that we need to be careful before enforcing a specific convention.
These conventions need to be documented in the dev guide, so that we can use it as reference and it would avoid us having to figure out what are the conventions to follow - and their justification.
Is it expected that lgtm analysis takes so long? It seems to be in the 30-45 min range?

Regarding making the history more readable and the change/review easier to follow, would it useful to have commit message more explicit, as for example: "lgtm improvement: change BaseException to Exception", etc. and one commit per type of changes?

jlaehne commented:
> This is useful but my impression is that we need to be careful before enforcing a specific convention. These conventions need to be documented in the dev guide, so that we can use it as reference and it would avoid us having to figure out what are the conventions to follow - and their justification.

Many issues like unreachable code or duplicate assignments of the same variable are not necessarily about conventions but show potential problems with the code. Sometimes, they also help to find a different issue with the code. Others like unused imports are not critical, but save overhead.
But indeed, some other things are more about conventions and there are also some false-positives, like the one I reported here: https://github.com/github/codeql/issues/9642#issuecomment-1162774212
So we might have to deactivate some types of reports.

LGTM-PR-reports just list newly introduced or resolved issues. I think that would be very helpful.

> Is it expected that lgtm analysis takes so long? It seems to be in the 30-45 min range?

I don't know. It was just an observation.

> Regarding making the history more readable and the change/review easier to follow, would it useful to have commit message more explicit, as for example: "lgtm improvement: change BaseException to Exception", etc. and one commit per type of changes?

I started out going by the order reported on the site and did not want to end up with too many commits, but when I realized that it could get messy I did indeed start grouping things like with the BaseException, also because that is a disputable convention. Too sort the other ones in grouped commits may be not so straightforward now.

jlaehne commented:
replaced by #2977
